File: src/backend/inference.js
const { spawn } = require('child_process');
const path = require('path');
const { getInferenceConfig } = require('./config');

async function runInference(modelId, message, preset) {
  try {
    const config = await getInferenceConfig();
    const scriptPath = path.join(__dirname, 'inference.py');
    const inputData = JSON.stringify({
      modelId,
      message,
      config: {
        ...config,
        temperature: preset === 'default' ? 0.7 : 0.5,
        top_p: 0.9,
        max_tokens: 512,
        model_dir: config.modelDir,
      },
    });

    return new Promise((resolve, reject) => {
      const pythonProcess = spawn(config.pythonPath, [scriptPath, inputData], {
        stdio: ['pipe', 'pipe', 'pipe'],
      });

      let output = '';
      let errorOutput = '';

      pythonProcess.stdout.on('data', (data) => {
        output += data.toString();
      });

      pythonProcess.stderr.on('data', (data) => {
        errorOutput += data.toString();
      });

      pythonProcess.on('close', (code) => {
        if (code !== 0) {
          reject(new Error(`Python process failed (code ${code}): ${errorOutput}`));
          return;
        }
        try {
          const result = JSON.parse(output);
          if (result.error) {
            reject(new Error(result.error));
          } else {
            resolve(result.response);
          }
        } catch (err) {
          reject(new Error(`Failed to parse Python output: ${err.message}\nOutput: ${output}\nError: ${errorOutput}`));
        }
      });

      pythonProcess.on('error', (err) => {
        reject(new Error(`Failed to start Python process: ${err.message}`));
      });
    });
  } catch (err) {
    throw new Error(`Inference setup failed: ${err.message}`);
  }
}

module.exports = { runInference };

File: src/backend/inference.py
# src/backend/inference.py
import json
import sys
from llama_cpp import Llama

def run_inference(model_id, message, config):
    try:
        model_path = f"{config['model_dir']}/{model_id}.gguf"
        llm = Llama(
            model_path=model_path,
            n_gpu_layers=config.get('n_gpu_layers', 0),
            n_threads=config.get('n_threads', 4),
            n_ctx=config.get('n_ctx', 4096),  # Match model training context
            n_batch=config.get('n_batch', 512),
        )
        params = {
            "max_tokens": 512,
            "temperature": 0.7,
            "top_p": 0.9,
        }
        response = llm(message, **params)
        return {"response": response["choices"][0]["text"].strip()}
    except Exception as e:
        return {"error": str(e)}

if __name__ == "__main__":
    try:
        input_data = json.loads(sys.argv[1])
        model_id = input_data["modelId"]
        message = input_data["message"]
        config = input_data["config"]
        result = run_inference(model_id, message, config)
        print(json.dumps(result))
    except Exception as e:
        print(json.dumps({"error": str(e)}))